File: trees/construct_bt_from_inorder_and_postorder.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {Integer[]} inorder
# @param {Integer[]} postorder
# @return {TreeNode}
def build_tree(inorder, postorder)
    return if inorder.nil? || postorder.nil? || inorder.empty? || postorder.empty?
    root = TreeNode.new(postorder.pop)

    idx = inorder.find_index(root.val)
    root.right = build_tree(inorder[idx + 1...inorder.size], postorder)
    root.left = build_tree(inorder[0...idx], postorder)
    
    root
end

File: trees/construct_bt_from_inorder_and_postorder.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def buildTree(self, inorder, postorder):
        if not inorder or not postorder: return
        root_val = postorder.pop()

        root = TreeNode(root_val)
        i = inorder.index(root_val)
        root.right = self.buildTree(inorder[i + 1:], postorder)
        root.left = self.buildTree(inorder[:i], postorder)
        return root


# 106. Construct Binary Tree from Inorder and Postorder Traversal
# https://leetcode.com/problems/construct-binary-tree-from-inorder-and-postorder-traversal/description/
